### ISSUE TITLE
Inconsistent text view font

### ISSUE THREAD

**Issue by AndrewBarba** [GitHawkApp/MessageViewController#49]
The latest release, 0.2.0, is causing the font to change in the text view as I type. Also introducing the left button has caused some really strange layout issues. For example, if I set `showsLeftButton` to false, the insets are rendered completely off center.

**Comment by rnystrom** [comment]
Blame #33

**Comment by rnystrom** [comment]
@AndrewBarba does it repro in the example if you just disable the left button?

**Comment by rnystrom** [comment]
@AndrewBarba I can't seem to repro this in the example. Can you give some more details? Are you also setting the left button inset?

**Comment by AndrewBarba** [comment]
Simply updating the library (no other code changes), went from this:

<img width="414" alt="screen shot 2018-03-16 at 5 39 20 pm" src="https://user-images.githubusercontent.com/1316152/37546027-6d66e95c-2941-11e8-8a6d-b175e42c5f64.png">

To this:

<img width="407" alt="screen shot 2018-03-16 at 5 41 52 pm" src="https://user-images.githubusercontent.com/1316152/37546032-73fe1786-2941-11e8-8b93-0fe17fc9a895.png">

And then typing notice how the first letter is a completely different font than the rest:

<img width="397" alt="screen shot 2018-03-16 at 5 42 09 pm" src="https://user-images.githubusercontent.com/1316152/37546045-814296d8-2941-11e8-86f4-8fa58a27e83e.png">

**Comment by AndrewBarba** [comment]
Okay this stuff for frame based on insets is all just completely wrong:

```swift
        // adjust by bottom offset so content is flush w/ text view
        let leftButtonFrame = CGRect(
            x: insetBounds.minX,
            y: (insetBounds.minY + textViewHeight) - leftButtonSize.height + leftButton.bottomHeightOffset,
            width: leftButtonSize.width,
            height: leftButtonSize.height
        )
        leftButton.frame = (showLeftButton) ? leftButtonFrame : .zero
        
        let textViewFrame = CGRect(
            x: ((showLeftButton) ? leftButtonFrame.maxX : 0) + leftButtonInset,
            y: insetBounds.minY,
            width: insetBounds.width - ((showLeftButton) ? leftButtonSize.width : 0) - leftButtonInset - rightButtonSize.width,
            height: textViewHeight
        )
        textView.frame = textViewFrame

        // adjust by bottom offset so content is flush w/ text view
        let rightButtonFrame = CGRect(
            x: textViewFrame.maxX + leftButtonInset,
            y: textViewFrame.maxY - rightButtonSize.height + rightButton.bottomHeightOffset,
            width: rightButtonSize.width,
            height: rightButtonSize.height
        )
        rightButton.frame = rightButtonFrame
```

**Comment by AndrewBarba** [comment]
Why not just use a stack view for this?

**Comment by rnystrom** [comment]
Ya fixing the inset stuff now. It's wrong.

> Why not just use a stack view for this?

Not going to involve AL layout hooks in this, I want fine control over what layout happens and when so that I can control when the text view resizes.

**Comment by AndrewBarba** [comment]
Yup makes sense. While you're fixing that, any chance you can decouple the font property from setting the font for text view and all buttons? I'd prefer to set the fonts individually

**Comment by rnystrom** [comment]
Agree, stripping that part

**Comment by AndrewBarba** [comment]
@rnystrom I'm still seeing really strange layout and font issues on 0.2.1.

On load:

<img width="401" alt="screen shot 2018-03-17 at 12 33 04 pm" src="https://user-images.githubusercontent.com/1316152/37557723-690ecea8-29df-11e8-8198-af5dfa414c0d.png">

After typing:

<img width="410" alt="screen shot 2018-03-17 at 12 33 14 pm" src="https://user-images.githubusercontent.com/1316152/37557725-6ee187ee-29df-11e8-964e-8d1c6579a6eb.png">


And here's my config:

```swift
        // Setup the button using text or an icon
        messageView.showLeftButton = false

        // Right button
        messageView.setButton(title: "Send".uppercased(), for: .normal, position: .right)
        messageView.setButton(font: UIFont(barstool: .branding, size: 14), position: .right)
        messageView.addButton(target: self, action: #selector(handleSendButtonTapped), position: .right)
        messageView.rightButtonTint = .barstoolBrightBlue

        // Change the appearance of the text view and its content
        messageView.font = UIFont(barstool: .branding, size: 14)
        messageView.textView.font = UIFont(barstool: .regular, size: 14)
        messageView.textView.placeholderText = "Type comment..."
        messageView.textView.placeholderTextColor = .lightGray
        messageView.textView.tintColor = .barstoolBrightBlue
        messageView.applyBorder(.top, color: .separator, size: 0.5)
        messageView.inset = UIEdgeInsets(top: 20, left: 16, bottom: 20, right: 16)

        // Capsule view
        let containerView = UIView()
        containerView.backgroundColor = UIColor(red: 0.945, green: 0.945, blue: 0.945, alpha: 1)
        containerView.borderColor = .separator
        containerView.borderWidth = 1
        containerView.cornerRadius = 3

        messageView.insertSubview(containerView, at: 0)
        containerView.snp.makeConstraints {
            $0.top.equalToSuperview().inset(8)
            $0.leading.equalToSuperview().inset(8)
            $0.trailing.equalToSuperview().inset(8)
            if #available(iOS 11, *) {
                $0.bottom.equalTo(messageView.safeAreaLayoutGuide.snp.bottom).inset(8)
            } else {
                $0.bottom.equalToSuperview().inset(8)
            }
        }
```

**Comment by rnystrom** [comment]
I’ll try to repro the font issue today and send another patch, I actually didn’t see this issue so didn’t make any change for it.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lenkaiser** [comment]
@rnystrom Thanks for the initial fix.

@AndrewBarba You can change your inset to:
```
messageView.inset = UIEdgeInsets(top: 20, left: 8, bottom: 20, right: 16)
```

Even when you're enabling the left button there is the correct space between left button and the textview. The original example uses this inset as well.

**Comment by AndrewBarba** [comment]
Okay thanks. Still need to figure out that initial right button position and the font issues. You can see the font issue in the screen you posted above. Is it the order I’m setting the fonts? Really strange

**Comment by lenkaiser** [comment]
@AndrewBarba I had the same problem. You can fix this by settings your inset before your settings the left- and right button properties. 

```
messageView.inset = UIEdgeInsets(top: 20, left: 8, bottom: 20, right: 16)
messageView.font = UIFont.preferredFont(forTextStyle: .body)

// Right button
messageView.setButton(title: "Send".uppercased(), for: .normal, position: .right)
messageView.setButton(font: UIFont.systemFont(ofSize: 14), position: .right)
messageView.addButton(target: self, action: #selector(onRightButton), position: .right)
messageView.rightButtonTint = .blue
```

If you set the inset afterwards you get weird behaviour indeed because the calculation rely on the insets.

**Comment by lenkaiser** [comment]
@rnystrom I’ll have a look at the font issue today. I think we need to update the readme because it uses a 16px inset. What do you think?

**Comment by rnystrom** [comment]
Seems like we need to lock off the exact same layout mechanisms for any of these property changes.

@lenkaiser not sure I follow about the readme

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lenkaiser** [comment]
@rnystrom Never mind, my bad. I was thinking the same code was used in the example but that works fine. 

I did find the problem, the textview uses an attributed string and uses different font sizes. The first character is the correct font but all the other after that use pointSize 17.0. If I input a simple text: Jj and print the attributes we can see the difference between the 2 letters. The reason why is unclear to me yet. Any suggestions?

```
attributes >> Optional(J{
    NSFont = "<UICTFont: 0x7fa416c1f850> font-family: \".SFUIText\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
}j{
    NSFont = "<UICTFont: 0x7fa416c17de0> font-family: \".SFUIText\"; font-weight: normal; font-style: normal; font-size: 17.00pt";
    "com.messageviewcontroller.autocompletekey" = 0;
})
```

**Comment by rnystrom** [comment]
@lenkaiser what’s the best way to repro the issue? If we can reproduce it in a failing unit test then fix the test that’d be ideal.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lenkaiser** [comment]
@rnystrom I'll work on that upcoming weekend.

**Comment by rnystrom** [comment]
Should be fixed in #59
